#40 implement a module page

PhilippeLeopoldie opened the pull request:
Implement student and teacher module pages with improved routing and navigation

This PR introduces dedicated module pages for students and teachers, refactors the shared Modules component, and updates navigation to support role-based access.



- Navigation
  - Updated page links with new routes:
   /courses/{courseId}/modules

 - Improved NavLink paths for teacher course navigation.

- Modules Component Refactor (ModuleComponent.razor)
   - Extracted a reusable Modules component.
   - Added better loading state and error handling.
   - Simplified initialization logic and cleaned up query string generation.

- Module Page (Modules.razor)
   - New role-specific page restricted to "Student" and "Teacher".
   - Displays modules per course.

- Other Updates
   - Updated module navigation to include links to activities by course/module ID.

PhilippeLeopoldie committed:
Update sidebar navigation for student and teacher roles

Changed student course link to /modules-student and added
new teacher link to /modules-teacher in Sidebar.razor.

PhilippeLeopoldie committed:
refactor(Modules): improve structure and logic

- Extract reusable module component
- Apply it to student and teacher modules pages

PhilippeLeopoldie committed:
Add student modules display component

This commit introduces a new Razor component for displaying modules specifically for students. It includes necessary using directives, sets up routing, and enforces authorization for users with the "Student" role. The component shows a loading message while fetching modules and displays them once loaded. Additionally, it defines parameters for `courseId` and `includeActivities`, and injects an API service to retrieve module data asynchronously during initialization.

PhilippeLeopoldie committed:
Add teacher module page component

This commit introduces a new Razor component for the teacher's module page. It includes necessary using directives, sets up routing with the `@page` directive, and enforces authorization for users with the "Teacher" role. The page displays a loading message while fetching modules and shows the modules once loaded. Additionally, it defines a `courseId` parameter, initializes a list of modules, and utilizes an API service to fetch module data asynchronously during initialization.

PhilippeLeopoldie committed:
Update includeActivities property with test comment

Modified the `includeActivities` property to include a comment indicating that the value is temporary for testing purposes.

PhilippeLeopoldie committed:
Add error handling for API calls in modules components

Introduce an `errorMessage` property in `ModulesStudent.razor` and `ModulesTeacher.razor` to capture and display error messages from API calls. Update the UI to show alerts for errors instead of just loading status. Wrap API calls in try-catch blocks for improved error handling and user feedback.

PhilippeLeopoldie committed:
merge(development): sync latest changes into #40-implement module page

PhilippeLeopoldie committed:
Update NavLink paths for teacher course navigation

Modified the `href` attribute of the `NavLink` component
to include a new path structure for teacher courses.
This change updates the routing to improve navigation
for teachers by directing them to
`/teacher/courses/{course.Id}/modules` instead of
the previous relative path.

PhilippeLeopoldie committed:
Improve error handling and loading state in Modules.razor

Updated `Modules.razor` to enhance error handling by displaying error messages during data fetching. Added a loading message while modules are being loaded. Refactored initialization logic for better loading state management and improved readability of query string generation. Changed `modules` property to private and removed unnecessary commented-out code.

PhilippeLeopoldie committed:
Update module routing and component structure

- Changed route from "/modules-student" to "/student/courses/{courseId:int}/modules".
- Updated page title from "Student" to "Elev".
- Modified module rendering to pass `courseId` to the `Modules` component.
- Removed `includeActivities` parameter; retained only `courseId`.
- Eliminated `OnInitializedAsync` method that fetched modules from the API.
- Retained `AuthorizeView` for role-based access control.

PhilippeLeopoldie committed:
Update ModulesTeacher component for course-specific routing

- Changed route from "/modules-teacher" to "/teacher/courses/{courseId:int}/modules".
- Updated page title from "Teacher" to "Lärare".
- Modified module rendering to use `courseId` instead of `modules`.
- Removed default value for `courseId` and eliminated `modules` property.
- Simplified initialization by removing error handling and loading logic.
- Deleted `OnInitializedAsync` method, removing the API call for modules.

PhilippeLeopoldie committed:
Update routing for ModulesStudent.razor

Changed the route from "/modules-student" to "/student/courses/{courseId:int}/modules" for better navigation. The authorization attribute for "Student" role remains intact to restrict access.

PhilippeLeopoldie committed:
Remove temporary comment from courseId property

Updated the `courseId` property in `ModulesStudent.razor` and `ModulesTeacher.razor` by removing the comment (`//temporary value for testing`). The property remains a public integer and retains the `[Parameter]` attribute.

PhilippeLeopoldie committed:
Update module navigation to activities

Removed activity count calculation and added a NavLink
for navigating to the activities page of each module,
using a URL format that includes the course ID and
module ID for better routing.

JorgenHagglund reviewed (approved):
Ser bra ut.
En fundering, dock...
Går det inte att slå samman ModulesStudent och ModulesTeacher med en AuthorizeView?

PhilippeLeopoldie commented:
> Ser bra ut. En fundering, dock... Går det inte att slå samman ModulesStudent och ModulesTeacher med en AuthorizeView?

Absolut, du har rätt, det skulle vara lättare.

PhilippeLeopoldie committed:
Refactor module navigation and management components

Updated navigation links and implemented role-based access
for module management in the Blazor application. Introduced
`ModulesComponent` for better encapsulation of module logic,
simplified student and teacher module views, and ensured
consistent routing in the sidebar.

ahlgren1234 reviewed (approved):
Ser bra ut.

JorgenHagglund reviewed (approved):
Ser bra ut!

ChornaIryna commented on LMS.Blazor/Components/Layout/Sidebar.razor:
```diff
@@ -16,6 +17,7 @@
         <AuthorizeView Roles="Teacher">
             <NavItem Href="/" Icon="bi-speedometer2" Text="�versikt" />
             <NavItem Href="/courses" Icon="bi-journal-bookmark" Text="Alla kurser" />
+            <NavItem Href="/modules" Icon="bi-list-ul" Text="Moduler" />
```
Länken till modulsidan bör inte finnas i lärarens meny, eftersom den kräver ett courseId för att fungera korrekt. Utan detta parameter leder länken till en ogiltig eller tom vy.

ChornaIryna commented on LMS.Blazor.Client/Pages/Modules/Modules.razor:
I Modules.razor används två separata <AuthorizeView>-block för "Teacher" och "Student", men båda renderar exakt samma komponent: <ModulesComponent courseId=@courseId />. Vad är mening  med rollspecifika block om logiken är identisk?

ChornaIryna commented on LMS.Blazor.Client/Pages/Modules/ModulesComponent.razor:
I ModulesComponent.razor visas knappen "Skapa modul" för alla användare, oavsett roll. Omslut knappen med <AuthorizeView Roles="Teacher"> för att säkerställa att endast lärare ser den.

ChornaIryna commented on LMS.Blazor/Components/Layout/Sidebar.razor:
```diff
@@ -8,6 +8,7 @@
     <nav class="flex-column">
         <AuthorizeView Roles="Student">
             <NavItem Href="/" Icon="bi-speedometer2" Text="�versikt" />
+            <NavItem Href="/modules-student" Icon="bi-list-ul" Text="Moduler" />
```
Fel länk. Ta bort eller ersätt länken med en dynamisk variant som inkluderar courseId.

ChornaIryna commented on LMS.Blazor.Client/Pages/Modules/Modules.razor:
```diff
@@ -0,0 +1,30 @@
+
+@page "/courses/{courseId:int}/modules"
+@rendermode @(new InteractiveWebAssemblyRenderMode(prerender: false))
+@* @attribute [Authorize(Roles ="Teacher")] *@
```
Rensa bort dessa om de inte används.
